File: src/tasks/cli/command/blockers/BlockersCommand.java
package tasks.cli.command.blockers;

import static omnia.data.cache.Memoized.memoize;
import static tasks.cli.command.Parameter.Repeatable.NOT_REPEATABLE;
import static tasks.cli.command.Parameter.Repeatable.REPEATABLE;

import omnia.data.cache.Memoized;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import tasks.cli.parser.ParserUtil;
import tasks.cli.command.Command;
import tasks.cli.command.FlagOption;
import tasks.cli.command.Option;
import tasks.cli.command.TaskOption;
import tasks.cli.command.TaskParameter;
import tasks.cli.parser.Parser;
import tasks.model.Task;

/** Canonical definition for the Blockers command. */
public final class BlockersCommand {
  private BlockersCommand() {}

  public static Command registration(
      Memoized<Parser<? extends List<ParserUtil.ParseResult<Task>>>> taskParser) {
    return Command.builder()
        .canonicalName("blockers")
        .aliases("blocker", "bk")
        .parameters(ImmutableList.of(new TaskParameter(NOT_REPEATABLE)))
        .options(OPTIONS.value())
        .parser(() -> new BlockersParser(taskParser))
        .helpDocumentation(
            "Modifies or lists blockers of an existing task. Can be used to add or remove blockers "
                + "from a task. If no modifications are specified, simply lists the existing "
                + "blockers for a task.");
  }

  static final Memoized<TaskOption> ADD_OPTION =
      memoize(
          () -> new TaskOption(
              "add",
              "a",
              "Adds another task as a blocker.",
              REPEATABLE));

  static final Memoized<FlagOption> CLEAR_OPTION =
      memoize(
          () -> new FlagOption(
              "clear",
              "c",
              "Removes all blocking tasks. Can be used together with --"
                  + ADD_OPTION.value().longName() + " to replace existing blockers with new ones.",
              NOT_REPEATABLE));

  static final Memoized<TaskOption> REMOVE_OPTION =
      memoize(
          () -> new TaskOption(
              "remove",
              "d",
              "Removes another task from being a blocker. Ignored if --"
                  + CLEAR_OPTION.value().longName() + " is set.",
              REPEATABLE));

  static final Memoized<ImmutableList<Option>> OPTIONS =
      memoize(
          () -> ImmutableList.of(
              ADD_OPTION.value(),
              CLEAR_OPTION.value(),
              REMOVE_OPTION.value()));
}

File: cli/src/me/strangepan/tasks/cli/feature/blockers/BlockersCommand.kt
package me.strangepan.tasks.cli.feature.blockers

import omnia.data.cache.Memoized
import omnia.data.cache.Memoized.Companion.memoize
import omnia.data.structure.immutable.ImmutableList
import me.strangepan.tasks.cli.command.Command
import me.strangepan.tasks.cli.command.FlagOption
import me.strangepan.tasks.cli.command.Parameter
import me.strangepan.tasks.cli.command.TaskOption
import me.strangepan.tasks.cli.command.TaskParameter

/** Canonical definition for the Blockers command.  */
object BlockersCommand {
  fun registration(): Command {
    return COMMAND.value()
  }

  val ADD_OPTION: Memoized<TaskOption> = memoize {
    TaskOption(
      "add",
      "a",
      "Adds another task as a blocker.",
      Parameter.Repeatable.REPEATABLE
    )
  }
  val CLEAR_OPTION: Memoized<FlagOption> = memoize {
    FlagOption(
      "clear",
      "c",
      "Removes all blocking me.strangepan.tasks.engine.tasks. Can be used together with " +
          "--${ADD_OPTION.value().longName()} to replace existing blockers with new ones.",
      Parameter.Repeatable.NOT_REPEATABLE
    )
  }
  val REMOVE_OPTION: Memoized<TaskOption> = memoize {
    TaskOption(
      "remove",
      "d",
      "Removes another task from being a blocker. Ignored if " +
          "--${CLEAR_OPTION.value().longName()} is set.",
      Parameter.Repeatable.REPEATABLE
    )
  }
  private val OPTIONS = memoize {
    ImmutableList.of(
        ADD_OPTION.value(),
        CLEAR_OPTION.value(),
        REMOVE_OPTION.value())
  }
  private val COMMAND: Memoized<Command> = memoize {
    Command.builder()
      .canonicalName("blockers")
      .aliases("blocker", "bk")
      .parameters(ImmutableList.of(TaskParameter(Parameter.Repeatable.REPEATABLE)))
      .options(OPTIONS.value())
      .helpDocumentation(
        "Modifies or lists blockers of an existing task. Can be used to add or remove "
            + "blockers from a task. If no modifications are specified, simply lists the "
            + "existing blockers for a task."
      )
  }
}